include/robot_interfaces/logger.hpp

```diff
@@ -0,0 +1,287 @@
+
+      header.push_back("#");
+      header.push_back("[Timestamp]");
+      header.push_back("(Current Index)");
+
+      for(i = 0; i < observation_name_.size(); i++)
+      {
+        for(j = 0; j < observation_data_[i].size(); j ++)
+        {
+          temp = observation_name_[i] + " " + std::to_string(j);
+          header.push_back(temp);
+        }
+      }
+
+      for(i = 0; i < applied_action_name_.size(); i++)
+      {
+        for(j = 0; j < applied_action_data_[i].size(); j ++)
+        {
+          temp = applied_action_name_[i] + " " + std::to_string(j);
+          header.push_back(temp);
+        }
+      }
+
+      for(i = 0; i < desired_action_name_.size(); i++)
+      {
+        for(j = 0; j < desired_action_data_[i].size(); j ++)
+        {
+          temp = desired_action_name_[i] + " " + std::to_string(j);
+          header.push_back(temp);
+        }
+      }
+
+      header.push_back("Status");
+
+    }
+
+    static void *write(void *instance_pointer)
+    {
+        ((FingerLogger *)(instance_pointer))->write();
+        return nullptr;
+    }
+
+    void write()
+    {
+
+        output_file_.open(output_file_name_);
+        std::ostream_iterator<std::string> output_iterator(output_file_, " , ");
+        get_header();
+        std::copy(header.begin(), header.end(), output_iterator);
+        output_file_ << std::endl;
+        output_file_.close();
+
+        while (!stop_was_called_ &&
+               !(logger_data_->desired_action->length() > 0))
+        {
+            real_time_tools::Timer::sleep_until_sec(0.1);
+        }
+
+        index_ = logger_data_->observation->newest_timeindex();
+
+        for (long int t = 0; !stop_was_called_; t++)
+        {
+            if (index_ + block_size_ ==
+                logger_data_->observation->newest_timeindex())
+            {
+                output_file_.open(output_file_name_, std::ios_base::app);
+
+
+#ifdef VERBOSE
+                auto t1 = std::chrono::high_resolution_clock::now();
+#endif
+
+                for (long int j = index_; j < index_ + block_size_; j++)
+                {
+                    try
+                    {
+                        std::ostream_iterator<double> output_iterator_(output_file_, " , ");
+                        applied_action = (*logger_data_->applied_action)[j];
+                        desired_action = (*logger_data_->desired_action)[j];
+                        observation = (*logger_data_->observation)[j];
+                        status = (*logger_data_->status)[j];
+
+                        observation_data_ = observation.get_data();
+                        applied_action_data_ = applied_action.get_data();
+                        desired_action_data_ = desired_action.get_data();
+
+                        output_file_ << logger_data_->observation->timestamp_s(j) << " , " << j << " , " ;
+
+                        for(loop_iterator_ = 0; loop_iterator_ < observation_data_.size(); loop_iterator_++)
```

Comment:
As above, the loop variable can be defined locally.  Also the name is a bit misleading as an "iterator" is a concept in C++ that is different from using a loop variable.  I think in this case it is totally fine to just call it `i`.

Same for the loops below.

Reply:
Actually, with modern C++ it might be possible to do the following:
```
for (auto data: observation_data_)
{
    std::copy(data.begin(), data.end(), output_iterator_);
}
```
However, I don't know if this is supported by the time series.  If you want, you can give it a try, but the current approach of using a loop variable is also totally fine.

Reply:
Regarding the auto method, I currently implemented this hack for dealing with the trailing comma for which I have to use the loop variable anyway. So will implement this when I find a better solution for the comma. Thanks :)